packages/core/src/components/essentials/color-scale-legend.ts

```diff
@@ -0,0 +1,327 @@
+import { DOMUtils } from '../../services';
+
+// D3 imports
+import { axisBottom } from 'd3-axis';
+import { scaleBand, scaleLinear } from 'd3-scale';
+import { interpolateRound, quantize } from 'd3-interpolate';
+
+export class ColorScaleLegend extends Legend {
+	type = 'color-legend';
+	renderType = RenderTypes.SVG;
+
+	private gradient_id =
+		'gradient-id-' + Math.floor(Math.random() * 99999999999);
+
+	init() {
+		const eventsFragment = this.services.events;
+
+		// Highlight correct circle on legend item hovers
+		eventsFragment.addEventListener(
+			Events.Axis.RENDER_COMPLETE,
+			this.handleAxisComplete
+		);
+	}
+
+	handleAxisComplete = (event: CustomEvent) => {
+		const svg = this.getComponentContainer();
+
+		const { width } = DOMUtils.getSVGElementSize(svg, {
+			useAttrs: true,
+		});
+
+		if (width > Configuration.legend.color.barWidth) {
+			const title = Tools.getProperty(
+				this.getOptions(),
+				'legend',
+				'colorLegend',
+				'title'
+			);
+
+			const { cartesianScales } = this.services;
+			// Get available chart area
+			const mainXScale = cartesianScales.getMainXScale();
+
+			const xDimensions = mainXScale.range();
+
+			// Align legend with the axis
+			if (xDimensions[0] > 1) {
+				svg.select('g.legend-rectangle').attr(
+					'transform',
+					`translate(${xDimensions[0]}, 0)`
+				);
+
+				if (title) {
+					const {
+						width: textWidth,
+					} = DOMUtils.getSVGElementSize(
+						svg.select('g.legend-title').select('text'),
+						{ useBBox: true }
+					);
+
+					// -9 since LEFT y-axis labels are moved towards the left by 9 by d3
+					const availableSpace = xDimensions[0] - textWidth - 9;
+
+					// If space is available align the the label with the axis labels
+					if (availableSpace > 1) {
+						svg.select('g.legend-title').attr(
+							'transform',
+							`translate(${availableSpace}, 0)`
+						);
+					} else {
+						// Move the legend down by 16 pixels to display legend text on top
+						svg.select('g.legend-rectangle').attr(
+							'transform',
+							`translate(${xDimensions[0]}, 16)`
+						);
+
+						// Align legend title with start of axis
+						svg.select('g.legend-title').attr(
+							'transform',
+							`translate(${xDimensions[0]}, 0)`
+						);
+					}
+				}
+			}
+		}
+	};
+
+	render(animate = false) {
+		const options = this.getOptions();
+
+		const customColors = Tools.getProperty(
+			options,
+			'color',
+			'gradient',
+			'colors'
+		);
+
+		const colorScaleType = Tools.getProperty(
+			options,
+			'legend',
+			'colorLegend',
+			'type'
+		);
+
+		let colorPairingOption = Tools.getProperty(
+			options,
+			'color',
+			'pairing',
+			'option'
+		);
+
+		const title = Tools.getProperty(
+			options,
+			'legend',
+			'colorLegend',
+			'title'
+		);
+
+		const customColorsEnabled = !Tools.isEmpty(customColors);
+		const domain = this.model.getValueDomain();
+
+		const svg = this.getComponentContainer();
+		svg.html('').attr('role', Roles.GROUP);
```

Comment:
so we're setting the SVG innerHTML to '' everytime? is there not a better way to render using d3 here?

Reply:
Fixed, switched over to AppendOrSelect for commonly used elements like title & axis and wrapper (<g>)